src/util/archive/fd_tar.c

```diff
@@ -0,0 +1,198 @@
+#include "fd_tar.h"
+#include "../fd_util.h"
+
+#include <errno.h>
+
+fd_tar_reader_t *
+fd_tar_reader_new( void *                       mem,
+                   fd_tar_read_vtable_t const * cb_vt,
+                   void *                       cb_arg ) {
+
+  if( FD_UNLIKELY( !mem ) ) {
+    FD_LOG_WARNING(( "NULL mem" ));
+    return NULL;
+  }
+  if( FD_UNLIKELY( !cb_vt || !cb_vt->file || !cb_vt->read ) ) {
+    FD_LOG_WARNING(( "NULL callback" ));
+    return NULL;
+  }
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)mem, fd_tar_reader_align() ) ) ) {
+    FD_LOG_WARNING(( "unaligned mem" ));
+    return NULL;
+  }
+
+  fd_tar_reader_t * self = (fd_tar_reader_t *)mem;
+  fd_memset( self, 0, sizeof(fd_tar_reader_t) );
+  self->cb_vt  = *cb_vt;
+  self->cb_arg = cb_arg;
+  return self;
+}
+
+void *
+fd_tar_reader_delete( fd_tar_reader_t * reader ) {
+  if( FD_UNLIKELY( !reader ) ) return NULL;
+  fd_memset( reader, 0, sizeof(fd_tar_reader_t) );
+  return reader;
+}
+
+static int
+fd_tar_process_hdr( fd_tar_reader_t * reader ) {
+
+  /* Detect EOF.  A TAR EOF is marked by 1024 bytes of zeros.
+     We abort after 512 bytes. */
+  int not_zero=0;
+  for( ulong i=0UL; i<sizeof(fd_tar_meta_t); i++ )
+    not_zero |= reader->buf[ i ];
+  if( !not_zero ) return -1;
+
+  fd_tar_meta_t const * hdr = (fd_tar_meta_t const *)reader->buf;
+
+  /* "ustar\x00" and "ustar  \x00" (overlaps with version) are both
+     valid values for magic.  These are POSIX ustar and OLDGNU versions
+     respectively. */
+  if( FD_UNLIKELY( 0!=memcmp( hdr->magic, FD_TAR_MAGIC, 5UL ) ) ) {
+    FD_LOG_WARNING(( "Invalid tar header magic at %#lx", reader->pos ));
+    FD_LOG_HEXDUMP_WARNING(( "Tar header", hdr, sizeof(fd_tar_meta_t) ));
+    return EPROTO;
+  }
+
+  ulong file_sz = fd_tar_meta_get_size( &reader->header );
+  if( FD_UNLIKELY( file_sz==ULONG_MAX ) ) {
+    FD_LOG_WARNING(( "Failed to parse file size in tar header" ));
+    return EPROTO;
+  }
+  reader->file_sz = file_sz;
+  reader->buf_ctr = (ushort)0U;
+
+  /* Ensure name is terminated */
+  reader->header.name[ FD_TAR_NAME_SZ-1 ] = '\0';
+
+  /* Call back to recipient */
+  int err = reader->cb_vt.file( reader->cb_arg, &reader->header, file_sz );
+  return fd_int_if( err, EIO, 0 );
+}
+
+static int
+fd_tar_read_hdr( fd_tar_reader_t * reader,
+                 uchar const **    pcur,
+                 uchar const *     end ) {
+
+  uchar const * cur = *pcur;
+
+  /* Skip padding */
+  ulong  pad_sz = fd_ulong_align_up( reader->pos, 512UL ) - reader->pos;
+         pad_sz = fd_ulong_min( pad_sz, (ulong)( end-cur ) );
+  cur += pad_sz;
+
+  /* Determine number of bytes to read */
+  long chunk_sz = (long)sizeof(fd_tar_meta_t) - (long)reader->buf_ctr;
+  FD_TEST( chunk_sz>=0L );
+  if( end-cur < chunk_sz ) chunk_sz = end-cur;
+
+  /* Copy to header */
+  fd_memcpy( reader->buf + reader->buf_ctr, cur, (ulong)chunk_sz );
+  cur             +=         chunk_sz;
+  reader->buf_ctr += (ushort)chunk_sz;
```

Comment:
FD_TEST this cast for paranoia (if someone changes fd_tar_meta_t in future or the compiler pads with junk)